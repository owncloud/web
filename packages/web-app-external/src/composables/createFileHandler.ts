import { Resource, SpaceResource, urlJoin } from '@ownclouders/web-client'
import { stringify } from 'qs'
import { useCapabilityStore, useClientService, useRequest } from '@ownclouders/web-pkg'

export const useCreateFileHandler = () => {
  const capabilityStore = useCapabilityStore()
  const clientService = useClientService()
  const { makeRequest } = useRequest({ clientService })

  const createFileHandler = async ({
    fileName,
    space,
    currentFolder
  }: {
    fileName: string
    space: SpaceResource
    currentFolder: Resource
  }) => {
    if (fileName === '') {
      return
    }

    const query = stringify({
      parent_container_id: currentFolder.fileId,
      filename: fileName
    })
    const url = `${capabilityStore.filesAppProviders[0].new_url}?${query}`
    const response = await makeRequest('POST', url)
    if (response.status !== 200) {
      throw new Error(`An error has occurred: ${response.status}`)
    }

    const path = urlJoin(currentFolder.path, fileName) || ''
    return clientService.webdav.getFileInfo(space, { path })
  }

  return {
    createFileHandler
  }
}
