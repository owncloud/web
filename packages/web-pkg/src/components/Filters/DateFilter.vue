<template>
  <div class="date-filter oc-flex" :class="`date-filter-${filterName}`">
    <oc-filter-chip
      ref="filterChip"
      :filter-label="filterLabel"
      :selected-item-names="selectedItem ? [selectedItem[displayNameAttribute]] : undefined"
      @clear-filter="clearFilter"
      @show-drop="onShowDrop"
      @hide-drop="onHideDrop"
    >
      <template #default>
        <oc-list class="date-filter-list" :class="{ 'date-filter-list-hidden': dateRangeClicked }">
          <li v-for="(item, index) in displayedItems" :key="index" class="oc-my-xs">
            <oc-button
              class="date-filter-list-item oc-flex oc-flex-between oc-flex-middle oc-width-1-1 oc-p-xs"
              :class="{
                'date-filter-list-item-active': isItemSelected(item)
              }"
              justify-content="space-between"
              appearance="raw"
              :data-testid="item[displayNameAttribute as keyof Item]"
              @click="toggleItemSelection(item)"
            >
              <div class="oc-flex oc-flex-middle oc-text-truncate">
                <div class="oc-text-truncate oc-ml-s">
                  <slot name="item" :item="item" />
                </div>
              </div>
              <div class="oc-flex">
                <oc-icon v-if="isItemSelected(item)" name="check" />
              </div>
            </oc-button>
          </li>
          <li class="oc-my-xs">
            <oc-button
              class="date-filter-list-item oc-flex oc-flex-between oc-flex-middle oc-width-1-1 oc-p-xs"
              :class="{
                'date-filter-list-item-active': dateRangeApplied
              }"
              justify-content="space-between"
              appearance="raw"
              data-testid="custom-date-range"
              @click="dateRangeClicked = true"
            >
              <div class="oc-flex oc-flex-middle oc-text-truncate">
                <div class="oc-text-truncate oc-ml-s">
                  <span v-text="$gettext('Custom date range')" />
                </div>
              </div>
              <div class="oc-flex">
                <oc-icon v-if="dateRangeApplied" name="check" />
              </div>
            </oc-button>
          </li>
        </oc-list>
        <div
          class="date-filter-range-panel oc-py-s"
          :class="{ 'date-filter-range-panel-active': dateRangeClicked }"
        >
          <div class="oc-flex oc-flex-middle oc-flex-between oc-mb-m">
            <oc-button
              appearance="raw"
              class="date-filter-range-panel-back"
              :aria-label="$gettext('Go back to filter options')"
              @click="dateRangeClicked = false"
            >
              <oc-icon name="arrow-left-s" fill-type="line" />
            </oc-button>
            <span v-text="$gettext('Custom date range')" />
            <oc-button
              appearance="raw"
              class="date-filter-range-panel-close"
              :aria-label="$gettext('Close filter')"
              @click="filterChip.hideDrop()"
            >
              <oc-icon name="close" />
            </oc-button>
          </div>
          <div class="oc-mt-s">
            <oc-datepicker
              :label="$gettext('From')"
              type="date"
              :is-clearable="true"
              :current-date="fromDate"
              @date-changed="(value) => setDateRangeDate(value, 'from')"
            />
            <oc-datepicker
              :label="$gettext('To')"
              type="date"
              :is-clearable="true"
              :current-date="toDate"
              :min-date="fromDate ? fromDate : undefined"
              @date-changed="(value) => setDateRangeDate(value, 'to')"
            />
          </div>
          <div class="date-filter-apply-btn">
            <oc-button
              appearance="outline"
              variation="passive"
              size="small"
              :disabled="!dateRangeValid"
              @click="applyDateRangeFilter"
            >
              {{ $gettext('Apply') }}
            </oc-button>
          </div>
        </div>
      </template>
    </oc-filter-chip>
  </div>
</template>

<script lang="ts" setup>
import { computed, onMounted, ref, unref } from 'vue'
import omit from 'lodash-es/omit'
import { useRoute, useRouteQuery, useRouter } from '../../composables'
import { formatDateFromDateTime } from '../../helpers'
import { queryItemAsString } from '../../composables/appDefaults'
import { DateTime } from 'luxon'
import { useGettext } from 'vue3-gettext'
import { OcFilterChip } from '@ownclouders/design-system/components'

type Item = Record<string, string>

interface Props {
  filterLabel: string
  filterName: string
  items: Item[]
  idAttribute?: string
  displayNameAttribute?: string
}
interface Emits {
  (e: 'selectionChange', value: Item): void
}
const {
  filterLabel,
  filterName,
  items,
  idAttribute = 'id',
  displayNameAttribute = 'name'
} = defineProps<Props>()
const emit = defineEmits<Emits>()
const router = useRouter()
const { current: currentLanguage } = useGettext()
const currentRoute = useRoute()
const selectedItem = ref<Item>()
const displayedItems = ref(items)
const fromDate = ref<DateTime>()
const toDate = ref<DateTime>()
const dateRangeClicked = ref(false)
const filterChip = ref<typeof OcFilterChip>()

const queryParam = `q_${filterName}`
const currentRouteQuery = useRouteQuery(queryParam)

const getId = (item: Item) => {
  return item[idAttribute as keyof Item]
}

const setRouteQuery = () => {
  return router.push({
    query: {
      ...omit(unref(currentRoute).query, [queryParam]),
      ...(unref(selectedItem) && {
        [queryParam]: getId(unref(selectedItem))
      })
    }
  })
}

const isItemSelected = (item: Item) => {
  return unref(selectedItem) && getId(unref(selectedItem)) === getId(item)
}

const resetDateRange = () => {
  fromDate.value = undefined
  toDate.value = undefined
}

const toggleItemSelection = async (item: Item) => {
  resetDateRange()
  if (isItemSelected(item)) {
    selectedItem.value = undefined
  } else {
    selectedItem.value = item
    unref(filterChip).hideDrop()
  }
  await setRouteQuery()
  emit('selectionChange', unref(selectedItem))
}

const clearFilter = () => {
  selectedItem.value = undefined
  dateRangeClicked.value = false
  resetDateRange()
  emit('selectionChange', unref(selectedItem))
  setRouteQuery()
}

const onShowDrop = () => {
  displayedItems.value = items
}

const onHideDrop = () => {
  dateRangeClicked.value = false
}

const setSelectedItemsBasedOnQuery = () => {
  const id = queryItemAsString(unref(currentRouteQuery))
  if (!id) {
    return
  }

  const selected = items.find((s) => getId(s) === id)
  if (selected) {
    selectedItem.value = selected
    return
  }

  if (unref(dateRangeApplied)) {
    const dateRange = id.replace('range:', '')
    const from = Number(dateRange.split(' - ')[0])
    const to = Number(dateRange.split(' - ')[1])
    fromDate.value = DateTime.fromMillis(from)
    toDate.value = DateTime.fromMillis(to)
    selectedItem.value = unref(dateRangeOption)
  }
}

const dateRangeApplied = computed(() =>
  queryItemAsString(unref(currentRouteQuery))?.startsWith('range:')
)

const dateRangeValid = computed(() => {
  if (!unref(fromDate) || !unref(toDate)) {
    return false
  }
  return unref(fromDate) <= unref(toDate)
})

const dateRangeOption = computed(() => {
  if (!unref(fromDate) || !unref(toDate)) {
    return null
  }

  const from = formatDateFromDateTime(unref(fromDate), currentLanguage, DateTime.DATE_SHORT)
  const to = formatDateFromDateTime(unref(toDate), currentLanguage, DateTime.DATE_SHORT)
  const fromDateMillis = unref(fromDate).toMillis()
  const toDateMillis = unref(toDate).toMillis()

  return {
    [idAttribute]: `range:${fromDateMillis} - ${toDateMillis}`,
    [displayNameAttribute]: `${from} - ${to}`
  }
})

const applyDateRangeFilter = async () => {
  selectedItem.value = unref(dateRangeOption)
  await setRouteQuery()
  unref(filterChip).hideDrop()
  emit('selectionChange', unref(selectedItem))
}

const setDateRangeDate = (
  { date, error }: { date: DateTime; error: Error },
  type: 'from' | 'to'
) => {
  if (error) {
    console.error(error)
    return
  }

  const prop = type === 'from' ? fromDate : toDate

  if (!date) {
    prop.value = undefined
    return
  }

  const formattedDate = type === 'from' ? date.startOf('day') : date.endOf('day')
  prop.value = formattedDate
}

defineExpose({ setSelectedItemsBasedOnQuery })

onMounted(() => {
  setSelectedItemsBasedOnQuery()
})
</script>

<style lang="scss">
.date-filter {
  overflow: hidden;

  &-list {
    li {
      &:first-child {
        margin-top: 0 !important;
      }
      &:last-child {
        margin-bottom: 0 !important;
      }
    }

    &-item {
      line-height: 1.5;
      gap: 8px;

      &:hover,
      &-active {
        background-color: var(--oc-color-background-hover) !important;
      }
    }

    &-hidden {
      min-height: 225px;
      visibility: hidden;
      transition: visibility 0.4s 0s;
    }
  }

  &-apply-btn {
    text-align: end;
  }

  &-range-panel {
    transform: translateX(100%);
    transition:
      transform 0.4s ease,
      visibility 0.4s 0s;
    visibility: hidden;
    position: absolute;
    width: calc(100% - var(--oc-space-medium));
    background: #fff;
    top: 0;
    color: var(--oc-color-swatch-passive-default);

    &-active {
      visibility: unset;
      transform: translateX(0);

      .oc-card {
        overflow: unset;
      }
    }
  }

  .oc-card {
    overflow: hidden;
  }

  .oc-date-picker label {
    font-size: var(--oc-font-size-small);
  }
}
</style>
