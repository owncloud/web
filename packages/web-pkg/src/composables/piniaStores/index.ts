export * from './apps'
export * from './auth'
export * from './capabilities'
export * from './clipboard'
export * from './config'
export * from './extensionRegistry'
export * from './messages'
export * from './modals'
export * from './resources'
export * from './shares'
export * from './spaces'
export * from './theme'
export * from './user'
export * from './webWorkers'
export * from './inviteTokensList'
export * from './app'
