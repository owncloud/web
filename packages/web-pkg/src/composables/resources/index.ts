export * from './useCanBeOpenedWithSecureView'
export * from './useGetResourceContext'
export * from './useLoadFileInfoById'
export * from './useResourceContents'
export * from './useResourcePath'
