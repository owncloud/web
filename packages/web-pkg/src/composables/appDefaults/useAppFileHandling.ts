import { unref } from 'vue'

import { Resource } from '@ownclouders/web-client'
import { MaybeRef } from '../../utils'
import { ClientService } from '../../services'
import { FileContext } from './types'
import { FileResource, SpaceResource } from '@ownclouders/web-client'
import { useClientService } from '../clientService'
import { ListFilesOptions } from '@ownclouders/web-client/webdav'
import { WebDAV } from '@ownclouders/web-client/webdav'
import { useCapabilityStore, useUserStore } from '../piniaStores'

interface AppFileHandlingOptions {
  clientService: ClientService
}

export type FileContentOptions = { responseType?: 'arraybuffer' | 'blob' | 'text' } & Record<
  string,
  any
>
export type UrlForResourceOptions = Omit<Parameters<WebDAV['getFileUrl']>[2], 'isUrlSigningEnabled'>

export interface AppFileHandlingResult {
  getUrlForResource(
    space: SpaceResource,
    resource: Resource,
    options?: UrlForResourceOptions
  ): Promise<string>
  revokeUrl(url: string): void
  getFileInfo(fileContext: MaybeRef<FileContext>, options?: ListFilesOptions): Promise<Resource>
  getFileContents(fileContext: MaybeRef<FileContext>, options?: FileContentOptions): Promise<any>
  putFileContents(
    fileContext: MaybeRef<FileContext>,
    putFileOptions: { content?: string } & Record<string, any>,
    parentFolderId?: string
  ): Promise<FileResource>
}

export function useAppFileHandling({
  clientService
}: AppFileHandlingOptions): AppFileHandlingResult {
  clientService = clientService || useClientService()
  const capabilityStore = useCapabilityStore()
  const userStore = useUserStore()

  const getUrlForResource = (
    space: SpaceResource,
    resource: Resource,
    options?: UrlForResourceOptions
  ) => {
    return clientService.webdav.getFileUrl(space, resource, {
      isUrlSigningEnabled: capabilityStore.supportUrlSigning,
      username: userStore.user?.onPremisesSamAccountName,
      ...options
    })
  }

  const revokeUrl = (url: string) => {
    return clientService.webdav.revokeUrl(url)
  }

  // TODO: support query parameters
  const getFileContents = (
    fileContext: MaybeRef<FileContext>,
    options: { responseType?: 'arraybuffer' | 'blob' | 'text' } & Record<string, any>
  ) => {
    return clientService.webdav.getFileContents(
      unref(unref(fileContext).space),
      {
        fileId: unref(unref(fileContext).itemId),
        path: unref(unref(fileContext).item)
      },
      {
        ...options
      }
    )
  }

  const getFileInfo = (
    fileContext: MaybeRef<FileContext>,
    options: ListFilesOptions = {}
  ): Promise<Resource> => {
    return clientService.webdav.getFileInfo(
      unref(unref(fileContext).space),
      {
        path: unref(unref(fileContext).item),
        fileId: unref(unref(fileContext).itemId)
      },
      options
    )
  }

  const putFileContents = (
    fileContext: MaybeRef<FileContext>,
    options: { content?: string } & Record<string, any>,
    parentFolderId?: string
  ) => {
    return clientService.webdav.putFileContents(unref(unref(fileContext).space), {
      path: unref(unref(fileContext).item),
      fileId: parentFolderId,
      fileName: unref(unref(fileContext).fileName),
      ...options
    })
  }

  return {
    getUrlForResource,
    revokeUrl,
    getFileContents,
    getFileInfo,
    putFileContents
  }
}
