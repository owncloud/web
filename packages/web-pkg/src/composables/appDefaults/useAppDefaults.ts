import { computed, unref, Ref } from 'vue'
import { useRouter, useRoute, useRouteParam } from '../router'
import { ClientService } from '../../services'
import { basename, dirname } from 'path'

import { FileContext } from './types'
import {
  useAppNavigation,
  AppNavigationResult,
  contextQueryToFileContextProps,
  contextRouteNameKey,
  queryItemAsString
} from './useAppNavigation'
import { useAppConfig, AppConfigResult } from './useAppConfig'
import { useAppFileHandling, AppFileHandlingResult } from './useAppFileHandling'
import { useAppFolderHandling, AppFolderHandlingResult } from './useAppFolderHandling'
import { useAppDocumentTitle } from './useAppDocumentTitle'
import { RequestResult, useRequest } from '../authContext'
import { useClientService } from '../clientService'
import { MaybeRef } from '../../utils'
import { useDriveResolver } from '../driveResolver'
import { urlJoin } from '@ownclouders/web-client'
import { useAppsStore, useAuthStore } from '../piniaStores'
import { storeToRefs } from 'pinia'

// TODO: this file/folder contains file/folder loading logic extracted from preview and drawio extensions
// Discussion how to progress from here can be found in this issue:
// https://github.com/owncloud/web/issues/3301

interface AppDefaultsOptions {
  applicationId: string
  applicationName?: MaybeRef<string>
  clientService?: ClientService
}

export type AppDefaultsResult = AppConfigResult &
  AppNavigationResult &
  AppFileHandlingResult &
  RequestResult &
  AppFolderHandlingResult & {
    isPublicLinkContext: Ref<boolean>
    currentFileContext: Ref<FileContext>
  }

export function useAppDefaults(options: AppDefaultsOptions): AppDefaultsResult {
  const router = useRouter()
  const appsStore = useAppsStore()
  const currentRoute = useRoute()
  const clientService = options.clientService ?? useClientService()
  const applicationId = options.applicationId

  const authStore = useAuthStore()
  const { publicLinkContextReady } = storeToRefs(authStore)

  const driveAliasAndItem = useRouteParam('driveAliasAndItem')
  const { space, item, itemId, loading } = useDriveResolver({ driveAliasAndItem })
  const currentFileContext = computed((): FileContext => {
    if (unref(loading)) {
      return null
    }
    let path: string
    if (unref(space)) {
      path = urlJoin(unref(space).webDavPath, unref(item))
    } else {
      // deprecated.
      path = urlJoin(queryItemAsString(unref(currentRoute).params.filePath))
    }

    return {
      path,
      driveAliasAndItem: unref(driveAliasAndItem),
      space: unref(space),
      item: unref(item),
      itemId: unref(itemId),
      fileName: basename(path),
      routeName: queryItemAsString(unref(currentRoute).query[contextRouteNameKey]),
      ...contextQueryToFileContextProps(unref(currentRoute).query)
    }
  })

  useAppDocumentTitle({
    appsStore,
    applicationId,
    applicationName: options.applicationName,
    currentFileContext,
    currentRoute
  })

  return {
    isPublicLinkContext: publicLinkContextReady,
    currentFileContext,
    ...useAppConfig({ appsStore, ...options }),
    ...useAppNavigation({ router, currentFileContext }),
    ...useAppFileHandling({
      clientService
    }),
    ...useAppFolderHandling({
      clientService,
      currentRoute
    }),
    ...useRequest({ clientService, currentRoute: unref(currentRoute) })
  }
}
