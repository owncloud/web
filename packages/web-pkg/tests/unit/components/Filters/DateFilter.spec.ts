import DateFilter from '../../../../src/components/Filters/DateFilter.vue'
import {
  PartialComponentProps,
  defaultComponentMocks,
  defaultPlugins,
  mount
} from '@ownclouders/web-test-helpers'
import { queryItemAsString } from '../../../../src/composables/appDefaults'
import { DateTime } from 'luxon'

vi.mock('../../../../src/composables/appDefaults')

const filterItems = [
  { id: '1', name: 'today' },
  { id: '2', name: 'yesterday' }
]

const selectors = {
  filterListItem: '.date-filter-list-item',
  filterChipLabel: '.oc-filter-chip-label',
  activeFilterListItemSpan: '.date-filter-list-item-active .oc-text-truncate span',
  customDateRangeBtn: '[data-testid="custom-date-range"]',
  customDateRangePanel: '.date-filter-range-panel',
  customDateRangeApplyBtn: '.date-filter-apply-btn button',
  customDateRangeBackBtn: '.date-filter-range-panel-back'
}

describe('DateFilter', () => {
  it('renders all items', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('can use a custom attribute as display name', () => {
    const filterItems = [
      { id: '1', displayName: 'today' },
      { id: '2', displayName: 'yesterday' }
    ]
    const { wrapper } = getWrapper({
      props: { displayNameAttribute: 'displayName', items: filterItems }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('route query', () => {
    it('sets the selected item as route query param', async () => {
      const { wrapper, mocks } = getWrapper()
      const item = wrapper.findAll(selectors.filterListItem).at(0)
      expect(mocks.$router.push).not.toHaveBeenCalled()
      await item.trigger('click')
      expect(mocks.$router.push).toHaveBeenCalledWith(
        expect.objectContaining({
          query: expect.objectContaining({ [(wrapper.vm as any).queryParam]: '1' })
        })
      )
    })
    it('sets the selected items initially when given via query param', () => {
      const { wrapper } = getWrapper({ initialQuery: '1' })
      expect((wrapper.vm as any).selectedItem).toEqual(filterItems[0])
    })
  })

  describe('custom date range', () => {
    it('shows the option and panel for picking a custom date range', () => {
      const { wrapper } = getWrapper()
      expect(wrapper.find(selectors.customDateRangeBtn).exists()).toBeTruthy()
      expect(wrapper.find(selectors.customDateRangePanel).exists()).toBeTruthy()
    })
    it('shows the custom date range as filter chip label when selected', async () => {
      const fromDate = DateTime.now().minus({ days: 1 })
      const toDate = DateTime.now()
      const { wrapper } = getWrapper({
        initialQuery: `range:${fromDate.toMillis()} - ${toDate.toMillis()}`
      })
      await wrapper.vm.$nextTick()
      const filterChipLabel = wrapper.find(selectors.filterChipLabel)
      expect(filterChipLabel.text()).toContain(
        `${fromDate.setLocale('en').toLocaleString()} - ${toDate.setLocale('en').toLocaleString()}`
      )
    })
    it('correctly marks the "Custom date range"-option as selected', async () => {
      const fromDate = DateTime.now().minus({ days: 1 })
      const toDate = DateTime.now()
      const { wrapper } = getWrapper({
        initialQuery: `range:${fromDate.toMillis()} - ${toDate.toMillis()}`
      })
      await wrapper.vm.$nextTick()
      expect(wrapper.find(selectors.activeFilterListItemSpan).text()).toEqual('Custom date range')
    })
    it('back button should close the custom date range panel', async () => {
      const { wrapper } = getWrapper()
      ;(wrapper.vm as any).dateRangeClicked = true
      await wrapper.vm.$nextTick()
      const backBtn = wrapper.find(selectors.customDateRangeBackBtn)
      await backBtn.trigger('click')
      expect((wrapper.vm as any).dateRangeClicked).toBeFalsy()
    })
    describe('apply button', () => {
      it('is not clickable without dates entered', async () => {
        console.error = vi.fn()
        const { wrapper } = getWrapper()
        const applyBtn = wrapper.find(selectors.customDateRangeApplyBtn)
        await applyBtn.trigger('click')
        expect(wrapper.emitted('selectionChange')).not.toBeDefined()
      })
      it('is not clickable when from date is after to date', async () => {
        const { wrapper } = getWrapper()
        ;(wrapper.vm as any).fromDate = DateTime.now().plus({ days: 1 })
        ;(wrapper.vm as any).toDate = DateTime.now()
        await wrapper.vm.$nextTick()
        const applyBtn = wrapper.find(selectors.customDateRangeApplyBtn)
        await applyBtn.trigger('click')
        expect(wrapper.emitted('selectionChange')).not.toBeDefined()
      })
      it('emits a selection change on click with today entered as date', async () => {
        const { wrapper } = getWrapper()
        ;(wrapper.vm as any).fromDate = DateTime.now()
        ;(wrapper.vm as any).toDate = DateTime.now()
        await wrapper.vm.$nextTick()
        const applyBtn = wrapper.find(selectors.customDateRangeApplyBtn)
        await applyBtn.trigger('click')
        expect(wrapper.emitted('selectionChange')).toBeDefined()
      })
    })
  })
})

function getWrapper({
  props = {},
  initialQuery = ''
}: { props?: PartialComponentProps<typeof DateFilter>; initialQuery?: string } = {}) {
  vi.mocked(queryItemAsString).mockImplementation(() => initialQuery)
  const mocks = defaultComponentMocks()

  return {
    mocks,
    wrapper: mount(DateFilter, {
      props: {
        filterLabel: 'Users',
        filterName: 'users',
        items: filterItems,
        ...props
      },
      slots: {
        item(data) {
          return props.displayNameAttribute ? data.item[props.displayNameAttribute] : data.item.name
        }
      },
      global: {
        plugins: [...defaultPlugins()],
        mocks,
        provide: mocks,
        stubs: { OcCheckbox: true }
      }
    })
  }
}
