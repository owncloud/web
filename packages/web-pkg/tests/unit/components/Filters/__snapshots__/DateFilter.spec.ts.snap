// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DateFilter > can use a custom attribute as display name 1`] = `
"<div class="date-filter oc-flex date-filter-users">
  <div class="oc-filter-chip oc-flex"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-filter-chip-button oc-pill" id="oc-filter-chip-5">
      <!--v-if-->
      <!-- @slot Content of the button --> <span class="oc-icon oc-icon-s oc-icon-passive oc-filter-check-icon-inactive"><!----></span> <span class="oc-text-truncate oc-filter-chip-label">Users</span>
      <!--v-if--> <span class="oc-icon oc-icon-s oc-icon-passive"><!----></span>
    </button>
    <div id="oc-drop-6" class="oc-drop oc-box-shadow-medium oc-rounded oc-filter-chip-drop">
      <!-- eslint-enable vuejs-accessibility/no-static-element-interactions -->
      <!-- eslint-enable vuejs-accessibility/click-events-have-key-events -->
      <div class="oc-card oc-card-body oc-background-secondary oc-p-s">
        <ul class="oc-list oc-my-rm oc-mx-rm date-filter-list">
          <li class="oc-my-xs"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-space-between oc-button-gap-m oc-button-passive oc-button-passive-raw date-filter-list-item oc-flex oc-flex-between oc-flex-middle oc-width-1-1 oc-p-xs" data-testid="today">
              <!--v-if-->
              <!-- @slot Content of the button -->
              <div class="oc-flex oc-flex-middle oc-text-truncate">
                <div class="oc-text-truncate oc-ml-s">today</div>
              </div>
              <div class="oc-flex">
                <!--v-if-->
              </div>
            </button></li>
          <li class="oc-my-xs"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-space-between oc-button-gap-m oc-button-passive oc-button-passive-raw date-filter-list-item oc-flex oc-flex-between oc-flex-middle oc-width-1-1 oc-p-xs" data-testid="yesterday">
              <!--v-if-->
              <!-- @slot Content of the button -->
              <div class="oc-flex oc-flex-middle oc-text-truncate">
                <div class="oc-text-truncate oc-ml-s">yesterday</div>
              </div>
              <div class="oc-flex">
                <!--v-if-->
              </div>
            </button></li>
          <li class="oc-my-xs"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-space-between oc-button-gap-m oc-button-passive oc-button-passive-raw date-filter-list-item oc-flex oc-flex-between oc-flex-middle oc-width-1-1 oc-p-xs" data-testid="custom-date-range">
              <!--v-if-->
              <!-- @slot Content of the button -->
              <div class="oc-flex oc-flex-middle oc-text-truncate">
                <div class="oc-text-truncate oc-ml-s"><span>Custom date range</span></div>
              </div>
              <div class="oc-flex">
                <!--v-if-->
              </div>
            </button></li>
        </ul>
        <div class="date-filter-range-panel oc-py-s">
          <div class="oc-flex oc-flex-middle oc-flex-between oc-mb-m"><button type="button" aria-label="Go back to filter options" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw date-filter-range-panel-back">
              <!--v-if-->
              <!-- @slot Content of the button --> <span class="oc-icon oc-icon-m oc-icon-passive"><!----></span>
            </button> <span>Custom date range</span> <button type="button" aria-label="Close filter" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw date-filter-range-panel-close">
              <!--v-if-->
              <!-- @slot Content of the button --> <span class="oc-icon oc-icon-m oc-icon-passive"><!----></span>
            </button></div>
          <div class="oc-mt-s">
            <div class="oc-date-picker"><label class="oc-label" for="oc-textinput-7">From</label>
              <div class="oc-position-relative">
                <!--v-if--> <input id="oc-textinput-7" aria-invalid="false" class="oc-text-input oc-input oc-rounded" type="date" value="">
                <!--v-if-->
              </div>
              <div class="oc-text-input-message">
                <!--v-if--> <span id="oc-textinput-7-message" class=""></span>
              </div>
              <!--v-if-->
            </div>
            <div class="oc-date-picker"><label class="oc-label" for="oc-textinput-8">To</label>
              <div class="oc-position-relative">
                <!--v-if--> <input id="oc-textinput-8" aria-invalid="false" class="oc-text-input oc-input oc-rounded" type="date" value="">
                <!--v-if-->
              </div>
              <div class="oc-text-input-message">
                <!--v-if--> <span id="oc-textinput-8-message" class=""></span>
              </div>
              <!--v-if-->
            </div>
          </div>
          <div class="date-filter-apply-btn"><button type="button" disabled="" class="oc-button oc-rounded oc-button-s oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-outline">
              <!--v-if-->
              <!-- @slot Content of the button --> Apply
            </button></div>
        </div>
      </div>
    </div>
    <!--v-if-->
  </div>
</div>"
`;

exports[`DateFilter > renders all items 1`] = `
"<div class="date-filter oc-flex date-filter-users">
  <div class="oc-filter-chip oc-flex"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-filter-chip-button oc-pill" id="oc-filter-chip-1">
      <!--v-if-->
      <!-- @slot Content of the button --> <span class="oc-icon oc-icon-s oc-icon-passive oc-filter-check-icon-inactive"><!----></span> <span class="oc-text-truncate oc-filter-chip-label">Users</span>
      <!--v-if--> <span class="oc-icon oc-icon-s oc-icon-passive"><!----></span>
    </button>
    <div id="oc-drop-2" class="oc-drop oc-box-shadow-medium oc-rounded oc-filter-chip-drop">
      <!-- eslint-enable vuejs-accessibility/no-static-element-interactions -->
      <!-- eslint-enable vuejs-accessibility/click-events-have-key-events -->
      <div class="oc-card oc-card-body oc-background-secondary oc-p-s">
        <ul class="oc-list oc-my-rm oc-mx-rm date-filter-list">
          <li class="oc-my-xs"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-space-between oc-button-gap-m oc-button-passive oc-button-passive-raw date-filter-list-item oc-flex oc-flex-between oc-flex-middle oc-width-1-1 oc-p-xs" data-testid="today">
              <!--v-if-->
              <!-- @slot Content of the button -->
              <div class="oc-flex oc-flex-middle oc-text-truncate">
                <div class="oc-text-truncate oc-ml-s">today</div>
              </div>
              <div class="oc-flex">
                <!--v-if-->
              </div>
            </button></li>
          <li class="oc-my-xs"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-space-between oc-button-gap-m oc-button-passive oc-button-passive-raw date-filter-list-item oc-flex oc-flex-between oc-flex-middle oc-width-1-1 oc-p-xs" data-testid="yesterday">
              <!--v-if-->
              <!-- @slot Content of the button -->
              <div class="oc-flex oc-flex-middle oc-text-truncate">
                <div class="oc-text-truncate oc-ml-s">yesterday</div>
              </div>
              <div class="oc-flex">
                <!--v-if-->
              </div>
            </button></li>
          <li class="oc-my-xs"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-space-between oc-button-gap-m oc-button-passive oc-button-passive-raw date-filter-list-item oc-flex oc-flex-between oc-flex-middle oc-width-1-1 oc-p-xs" data-testid="custom-date-range">
              <!--v-if-->
              <!-- @slot Content of the button -->
              <div class="oc-flex oc-flex-middle oc-text-truncate">
                <div class="oc-text-truncate oc-ml-s"><span>Custom date range</span></div>
              </div>
              <div class="oc-flex">
                <!--v-if-->
              </div>
            </button></li>
        </ul>
        <div class="date-filter-range-panel oc-py-s">
          <div class="oc-flex oc-flex-middle oc-flex-between oc-mb-m"><button type="button" aria-label="Go back to filter options" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw date-filter-range-panel-back">
              <!--v-if-->
              <!-- @slot Content of the button --> <span class="oc-icon oc-icon-m oc-icon-passive"><!----></span>
            </button> <span>Custom date range</span> <button type="button" aria-label="Close filter" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw date-filter-range-panel-close">
              <!--v-if-->
              <!-- @slot Content of the button --> <span class="oc-icon oc-icon-m oc-icon-passive"><!----></span>
            </button></div>
          <div class="oc-mt-s">
            <div class="oc-date-picker"><label class="oc-label" for="oc-textinput-3">From</label>
              <div class="oc-position-relative">
                <!--v-if--> <input id="oc-textinput-3" aria-invalid="false" class="oc-text-input oc-input oc-rounded" type="date" value="">
                <!--v-if-->
              </div>
              <div class="oc-text-input-message">
                <!--v-if--> <span id="oc-textinput-3-message" class=""></span>
              </div>
              <!--v-if-->
            </div>
            <div class="oc-date-picker"><label class="oc-label" for="oc-textinput-4">To</label>
              <div class="oc-position-relative">
                <!--v-if--> <input id="oc-textinput-4" aria-invalid="false" class="oc-text-input oc-input oc-rounded" type="date" value="">
                <!--v-if-->
              </div>
              <div class="oc-text-input-message">
                <!--v-if--> <span id="oc-textinput-4-message" class=""></span>
              </div>
              <!--v-if-->
            </div>
          </div>
          <div class="date-filter-apply-btn"><button type="button" disabled="" class="oc-button oc-rounded oc-button-s oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-outline">
              <!--v-if-->
              <!-- @slot Content of the button --> Apply
            </button></div>
        </div>
      </div>
    </div>
    <!--v-if-->
  </div>
</div>"
`;
